In [1]:
# Import necessary libraries and modules
from tensorflow.keras.layers import Dense, Input, Flatten, Conv2D, MaxPool2D #Max pooling operation for 2D spatial data.
from tensorflow.keras.models import Model
import matplotlib.pyplot as plt

In [ ]:
# defining a model named build_model
def build_model():
    # input layers with channel 1 for 28x28 grayscale image
	inputs = Input((28, 28, 1))
    # a convolution layer with 10 kernel of size 3x3, strides size 2x2 and valid padding means no padding used here
	x = Conv2D(10, (3,3), strides=(2,2), padding = 'valid')(inputs)
    # another convolution layer with 4 kernel of size 2x2, strides size 2x2 and valid padding means no padding used here
	x = Conv2D(4, (2,2), strides=(2,2), padding = 'valid')(x)
    # Flatten layer by flattening the previous layer
	x = Flatten()(x)
    # Dense layer with 100 neuron
	x = Dense(100, activation = 'relu')(x)
    # output layer with 10 class
	outputs = Dense(10, activation='softmax')(x)
    # model feeded with inputs and outputs
	model = Model(inputs, outputs)
    # for model summary
	model.summary()
    # for determining the loss of model
	model.compile(loss = 'mse', optimizer='rmsprop')
    
	return model

In [ ]:
def main():
	model = build_model()

In [ ]:
if __name__ == "__main__":
	main()

inputs = Input((28, 28, 1))
x = Conv2D(10, (3,3), strides=(2,2), padding = 'valid')(inputs)
x = Conv2D(4, (2,2), strides=(2,2), padding = 'valid')(x)
x = Flatten()(x)
x = Dense(100, activation = 'relu')(x)
outputs = Dense(10, activation='softmax')(x)


### VALID Padding: it means no padding and it assumes that all the dimensions are valid so that the input image gets fully covered by a filter and the stride specified by you.
### The padding type is called SAME because the output size is the same as the input size(when stride=1). Using 'SAME' ensures that the filter is applied to all the elements of the input. Normally, padding is set to "SAME" while training the model. Output size is mathematically convenient for further computation.

In [3]:
model = Model(inputs, outputs)
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 28, 28, 1)]       0         
                                                                 
 conv2d (Conv2D)             (None, 13, 13, 10)        100       
                                                                 
 conv2d_1 (Conv2D)           (None, 6, 6, 4)           164       
                                                                 
 flatten (Flatten)           (None, 144)               0         
                                                                 
 dense (Dense)               (None, 100)               14500     
                                                                 
 dense_1 (Dense)             (None, 10)                1010      
                                                                 
Total params: 15,774
Trainable params: 15,774
Non-trainable p

In [8]:
model.compile(loss = 'mse', optimizer='rmsprop')

#### Number of parameters in Keras Conv2D layer is calculated using the following equation:

#### number_parameters = out_channels * (in_channels * kernel_h * kernel_w + 1)  # 1 for bias
#### here, in_channels=1, out_channels=10,kernel_h=3,kernel_w=3 i.e 100
#### out_channels=4,inchannels=10,kernel_h=2,kernel_w=2 i.e 164
#### (144+1=145*100=14500, 100+1=101*10=1010